pkg/securitycontext/admission/admission.go

```diff
@@ -0,0 +1,177 @@
+	//     2.  set the uid on the SCC
+	// 3.  Match the pod's SC to an SCC; for each SCC:
+	//     1.  generate the SC
+	//     2.  validate against the pod's SC
+
+	if a.GetResource() != "pods" {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(kapi.Pod)
+	if !ok {
+		return errors.NewBadRequest("a pod was received, but could not convert the request object.")
+	}
+
+	ns := a.GetNamespace()
+	matchedConstraints, err := getMatchingSecurityContextConstraints(c.client, ns, pod)
+	if err != nil {
+		return err
+	}
+
+	for _, constraint := range matchedConstraints {
+		if constraint.RunAsUser.Type == kapi.RunAsUserStrategyMustRunAs && constraint.RunAsUser.UID == nil {
+			uid, err := getPreallocatedUID(constraint)
+			if err != nil {
+				return err
+			}
+
+			constraint.RunAsUser.UID = &uid
+		}
+
+		if constraint.SELinuxContext.Type == kapi.SELinuxStrategyMustRunAs && constraint.SELinuxContext.SELinuxOptions.Level == "" {
+			level, err := getPreallocatedLevel(constraint)
+			if err != nil {
+				return err
+			}
+
+			constraint.SELinuxContext.SELinuxOptions.Level = level
+		}
+	}
+
+	providers := make([]SecurityContextConstraintsProvider)
+
+	for _, constraint := range matchedConstraints {
+		provider, err := securitycontextconstraints.NewSimpleProvider(constraint, c.client)
+		if err != nil {
+			return err
+		}
+
+		providers = append(providers, provider)
+	}
+
+Containers:
+	for i, container := range pod.Spec.Containers {
+		for _, provider := range providers {
+			// Create a security context for this provider
+			context, err := provider.CreateSecurityContext(pod, container, nil)
+
+			// Create a deep copy of the pod to test with the security context
+			podCopyObj, err := kapi.Scheme.Copy(pod)
+			if err != nil {
+				return err
+			}
+			podCopy, ok := podCopyObj.(kapi.Pod)
+			if !ok {
+				// TODO: better error message
+				return errors.NewBadRequest("pod copy failed")
+			}
+			podCopy.Spec.Containers[i].SecurityContext = context
+
+			// Validate the copied pod against the security context provider
+			errs := provider.ValidateSecurityContext(podCopy, podCopy.Containers[i], nil)
+			if len(errs) != 0 {
+				continue
+			}
+
+			// Set the security context on containers
+			container.SecurityContext = context
+			continue Containers
+		}
+
+		// If we have reached this code, we couldn't find an SCC that matched
+		// the requested security context for this container
+		//
+		// TODO: better error message
+		return errors.NewBadRequest("Unable to find a valid security context constraint for container")
+	}
+
+	return nil
+}
+
+func getPreallocatedUID(constraint api.SecurityContextConstraint) (types.UID, error) {
+	return types.UID("1001"), nil
+}
+
+func getPreallocatedLevel(constraint api.SecurityContextConstraint) (string, error) {
+	return "s0", nil
+}
+
+func getMatchingSecurityContextConstraints(client client.Interface, ns string, pod *kapi.Pod) ([]kapi.SecurityContextConstraint, error) {
```

Comment:
Let's return a set or map here to avoid duplicate constraints (and therefore duplicate providers and duplicate runs of gen/validate against the same constraints)